# Assignment: Data Wrangling
### `! git clone https://github.com/ds4e/wrangling`
### Do Q1 and Q2, and either Q3 or Q4, for a total of 3 questions.

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [3]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt

Matplotlib is building the font cache; this may take a moment.


In [26]:
# 1. Missing Values (Numeric)
bnb = pd.read_csv('/Users/gracesaunders/Documents/GitHub/wrangling/assignment/data/airbnb_hw.csv')

In [28]:
"""
Cleaning the Price variable: removing commas and converting to numeric, then sorting to check that the high and low values look correct.
Remaining missing: 0
Cell generated by Data Wrangler. 
"""
def clean_data(df):
    df['Clean Price'] = df['Price'].str.replace(',', '') #remove commas from Price column
    df['Clean Price'] = pd.to_numeric(df['Clean Price']) #convert Price column to numeric
    df = df.sort_values(['Clean Price'], ascending=[False]) # Sort by column: 'Clean Price' (descending) 
    return df

bnb_clean = clean_data(bnb.copy())
bnb_clean.head()

Host Id  Host Since                                 Name  \
23927  23248648   11/2/2014       NO LONGER BOOKING RESERVATIONS   
4002    1177497   9/18/2011                        Film Location   
17900  11598359   1/24/2014     SUPER BOWL Brooklyn Duplex Apt!!   
2405     496780    4/9/2011       test listing which is not real   
24136  23564279  11/10/2014  Breathtaking 2-Story Soho Penthouse   

      Neighbourhood  Property Type  Review Scores Rating (bin)  \
23927      Manhattan     Apartment                        95.0   
4002        Brooklyn         House                         NaN   
17900       Brooklyn     Apartment                         NaN   
2405        Brooklyn     Apartment                         NaN   
24136      Manhattan     Apartment                         NaN   

             Room Type  Zipcode  Beds  Number of Records  Number Of Reviews  \
23927  Entire home/apt  10025.0   2.0                  1                 21   
4002   Entire home/apt  11205.0   5.0                  1                  0   
17900  Entire home/apt  11238.0   4.0                  1                  0   
2405   Entire home/apt  11249.0   5.0                  1                  0   
24136  Entire home/apt  10013.0   3.0                  1                  0   

        Price  Review Scores Rating  Clean Price  
23927  10,000                  95.0        10000  
4002    8,000                   NaN         8000  
17900   6,500                   NaN         6500  
2405    5,999                   NaN         5999  
24136   5,000                   NaN         5000

In [ ]:
#2: Missing Data (Categorical)
mn = pd.read_csv('/Users/gracesaunders/Documents/GitHub/wrangling/assignment/data/mn_police_use_of_force.csv')

In [33]:
mn['injury_NA'] = mn['subject_injury'].isnull() # Create an injured missing dummy; 

print("# of missing values for injury: ", sum(mn['injury_NA']),'\n') # How many missing values are there?
print("percent of missing values for injury: ",sum(mn['injury_NA'])/len(mn['injury_NA'])*100,'\n') # What proportion of the data is missing?

pd.crosstab(mn['force_type_action'], mn['force_type']) # Create a crosstab of force_type and force_type_action to look at specifics 

# of missing values for injury:  0 

percent of missing values for injury:  0.0 



force_type                                         Baton  Bodily Force  \
force_type_action                                                        
Arcing                                                 0             0   
Bean Bag Round                                         0             0   
Bicycle                                                0             0   
Body Weight Pin                                        0          2651   
Body Weight to Pin                                     0          1711   
Chemical Munitions                                     0             0   
Compliance                                             0             0   
Conscious Neck Restraint                               0            92   
Crowd Control Mace                                     0             0   
Crowd Control Techniques                               1             0   
Display                                                0             0   
Firearm as striking tool                               0             0   
Firearms As Striking Tool                              0             0   
Firing Darts                                           0             0   
Flashlight                                             0             0   
Hand Gun                                               0             0   
Jabs                                                   1             0   
Joint Lock                                             0          1578   
Kicks                                                  0           130   
Knee                                                   0           203   
Knees                                                  0           208   
Neck Restraint: subject did not lose conciousness      0            79   
Neck Restraint: subject lost conciousness              0            14   
Not Side Recovery Position                             0             0   
Off Lead                                               0             0   
On Lead                                                0             0   
Other (specify)                                        0             0   
Other Improvised Weapon                                0             0   
Personal Mace                                          0             0   
Punch                                                  0           231   
Punches                                                0           288   
Push Away                                              0           828   
Red Dot                                                0             0   
Side Recovery Position                                 0             0   
Slap                                                   0            31   
Slaps                                                  0            31   
Strikes                                                2             0   
Takedown                                               0          1330   
Touch                                                  0             0   
Unconscious Neck Restraint                             0            25   
Vehicle                                                0             0   

force_type                                         Chemical Irritant  Firearm  \
force_type_action                                                               
Arcing                                                             0        0   
Bean Bag Round                                                     0        0   
Bicycle                                                            0        0   
Body Weight Pin                                                    0        0   
Body Weight to Pin                                                 0        0   
Chemical Munitions                                                25        0   
Compliance                                                         0        0   
Conscious Neck Restraint                                           0        0   
Crowd C

In [ ]:

mn = mn.fillna({'subject_injury': "Unknown"}) # Fill missing values with "Unknown"
pd.crosstab(mn['subject_injury'], mn['force_type']) # Create a crosstab of subject_injury and force_type

force_type      Baton  Bodily Force  Chemical Irritant  Firearm  \
subject_injury                                                    
No                  0          1093                131        2   
Unknown             2          7051               1421        0   
Yes                 2          1286                 41        0   

force_type      Gun Point Display  Improvised Weapon  Less Lethal  \
subject_injury                                                      
No                             33                 34            0   
Unknown                        27                 74           87   
Yes                            44                 40            0   

force_type      Less Lethal Projectile  Maximal Restraint Technique  \
subject_injury                                                        
No                                   1                            0   
Unknown                              0                          170   
Yes                                  2                            0   

force_type      Police K9 Bite  Taser  
subject_injury                         
No                           2    150  
Unknown                     31    985  
Yes                         44    172

Observations from crosstab:

For incidents where force_type = Bodily Force, 7,051/9,430 rows are missing for subject_injured (this accounts for most of the missing data, since it's a very large majority of the largest category of force type)

It's difficult to conclude whether an injury resulted from all of these missing observations because 'Bodily Force' is such a broad range.

The second largest contributor to the missing data is force_type = Chemical Irritant (mostly mace), with 1,421 observations missing for subject_injured. 


Cleaning choices: 

Because so much critical information was missing from the data set, I chose to replace Missing values for subject_injured with 'Unknown,' because that seemed the most reflective of reality. 

It's tempting to 'fix' the data by guessing or assuming bassed on force_type, but I can't think of any 'objective' or 'accurate' way to do so. I don't think a boolean variable was an informative way to store information on injuries, because the difference between irritation from mace or tear gas is quite different from a dog bite or a 'maximal restraint technique.' For example, I found it particularly interesting that all observations for Maximal Restraint Technique were missing for subject_injured, considering that MRT was outlawed in MN after the murder of George Floyd. I wouldn't consider his injury status to be Unknown. 

However, with the given data structure, 'Unknown' seemed to be the most applicable, and I didn't feel that it was right to impute values based on my own guesses. 

In [81]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
pt = pd.read_csv(url,low_memory=False) 

In [50]:
#3. Dummy variable: clean `WhetherDefendantWasReleasedPretrial` 
pt.head()
var = 'WhetherDefendantWasReleasedPretrial' # variable to be cleaned
null_count = sum(pt[var].isnull()) # count how many missing values there are
print ("# of missing values for", var, ":", null_count, '\n') # there are zero null values, but the code book indicates that 9 is 'Unclear'

pt[var] = pt[var].replace( [9], np.nan) # replacing the code for 'Unclear' with NaN

print(pt[var].value_counts(), '\n') # examine counts of each unique value


# of missing values for WhetherDefendantWasReleasedPretrial : 0 

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64 



Codebook on Sentence type

Probation/no incarceration  0 -> 0
Jail up to 12 months 1 
Prison (one year or more) 2 
Other, all charges in cohort contact resulted in pending, dismissed, deferred, etc.  4 -> 0
Not applicable (e.g., disposition record not found, or unknown) 9

In [82]:
# 4. Missing values, not at random: 
sl = 'ImposedSentenceAllChargeInContactEvent' # variable to be cleaned - Total Imposed Sentence (months)
st = 'SentenceTypeAllChargesAtConvictionInContactEvent' # related - Sentence Type

pt[sl] = pd.to_numeric(pt[sl], errors='coerce') # coerce to numeric

pt['sl_na'] = pt[sl].isnull() # Create a missing dummy for sentence length

print("How many missing values for sentence length?")
print(sum(pt['sl_na']),'\n')

print("Cross tab missing sentence length with sentence type: \n")
print(pd.crosstab(pt['sl_na'],pt[st]),'\n') # cross tab missing sentence length with sentence type
# Shows that all of the missing values for sentence length are for sentence type 'Other' or 'Not Applicable'
# Logically, if the sentence type is 'Other' or 'Not Applicable', 
# they received no incarceration time so there should be no sentence length (i.e., 0 months).

#Replacing missing values with 0 where sentence type logically implies no sentence length
pt.loc[ pt[st] == 4, sl] = 0 # Sentence type is 'Other'
pt.loc[ pt[st] == 9, sl] = 0 # Sentence type is 'Not Applicable'

print("How many missing values are left?")
print(sum(pt['sl_na']),'\n') 



How many missing values for sentence length?
9053 

Cross tab missing sentence length with sentence type: 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
sl_na                                                                       
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 

How many missing values are left?
9053 



**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.


In [5]:
sa = pd.read_excel('/Users/gracesaunders/Documents/GitHub/wrangling/data/GSAF5.xls')

2. Drop any columns that do not contain data.


In [6]:
sa = sa.drop(columns=['Unnamed: 22'])
sa = sa.drop(columns=['Unnamed: 21'])

sa.head()

Date    Year        Type        Country  \
0  2025-01-11 00:00:00  2025.0    Provoked            USA   
1  2025-01-02 00:00:00  2025.0  Unprovoked  New Caledonia   
2  2025-01-02 00:00:00  2025.0  Unprovoked      Australia   
3  2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
4  2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   

                 State                                 Location      Activity  \
0               Hawaii            Off Haleiwa Boat Harbour Oahu        Diving   
1         Grande Terre              Islet of Kendek near Koumac  Spearfishing   
2      South Australia  Granites Beach near Westall Streaky Bay       Surfing   
3  North of Marsa Alam                                  Red Sea  SCUBA Diving   
4  North of Marsa Alam                                  Red Sea  SCUBA Diving   

                                     Name Sex Age  ... Fatal Y/N     Time  \
0  Male not stated was a dive tour worker   M  23  ...         N  1340hrs   
1                        Robert Cuewapuru   M  40  ...         Y   1615hr   
2                           Lance Appleby   M  28  ...         Y   1710hr   
3                         Peppino Fappani   M  69  ...         N        ?   
4                       Gianluca Di Gioia   M  48  ...         Y        ?   

                                 Species   \
0                              Not stated   
1          Reportedly Tiger or Bull shark   
2  Great White Shart estimated 4.8m 16 ft   
3                Reportedly a Tiger Shark   
4                Reportedly a Tiger Shark   

                                              Source  pdf href formula href  \
0                  Kevin McMurray Trackingsharks.com  NaN          NaN  NaN   
1                      Johannes Marchand Todd Smith   NaN          NaN  NaN   
2  Glen Folkard: Simon De Marchi News.com.au: The...  NaN          NaN  NaN   
3  Todd Smith : Kevin McMurray Trackingsharks .co...  NaN          NaN  NaN   
4  Todd Smith : Kevin McMurray Trackingsharks .co...  NaN          NaN  NaN   

  Case Number Case Number.1 original order  
0         NaN           NaN            NaN  
1         NaN           NaN            NaN  
2         NaN           NaN            NaN  
3         NaN           NaN            NaN  
4         NaN           NaN            NaN  

[5 rows x 21 columns]

3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?


In [12]:
# find range for Year column
print(sa['Year'].min(), "to ", sa['Year'].max())

0.0 to  2026.0


In [13]:
sa['Year'] = sa['Year'].replace(0, np.nan) # replacing zeroes with NaN
sa['Year'] = sa['Year'].replace(2026, 2024) # replace 2026 (future) with correct Year based on Date column

In [17]:
sa['Year'].describe()

count    6859.000000
mean     1971.384750
std        60.722703
min         5.000000
25%      1950.000000
50%      1987.000000
75%      2010.000000
max      2025.000000
Name: Year, dtype: float64

4. Clean the Age variable and make a histogram of the ages of the victims.

5. What proportion of victims are male?


6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?


7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.


8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?


9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

**Q3.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q4.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?